src/lib/integration_test_cloud_engine/kubernetes_network.ml

```diff
@@ -22,60 +25,91 @@ module Node = struct
     let kubectl_cmd =
-      Printf.sprintf "%s logs -c %s -n %s %s" base_kube_cmd node.container_id
+      Printf.sprintf "%s logs -c %s -n %s %s" base_kube_cmd container_id
         node.namespace pod
     in
     Util.run_cmd_exn cwd "sh" [ "-c"; kubectl_cmd ]
 
-  let run_in_container node cmd =
+  let run_in_container node ~container_id ~cmd =
     let base_args = base_kube_args node in
     let base_kube_cmd = "kubectl " ^ String.concat ~sep:" " base_args in
     let kubectl_cmd =
       Printf.sprintf
         "%s -c %s exec -i $( %s get pod -l \"app=%s\" -o name) -- %s"
-        base_kube_cmd node.container_id base_kube_cmd node.pod_id cmd
+        base_kube_cmd container_id base_kube_cmd node.pod_id cmd
     in
     let%bind.Deferred cwd = Unix.getcwd () in
-    Malleable_error.return (Util.run_cmd_exn cwd "sh" [ "-c"; kubectl_cmd ])
+    Util.run_cmd_exn cwd "sh" [ "-c"; kubectl_cmd ]
+
+  (* we dont use this functionality atm but if we need logic that runs commands in all containers a whole pod at a time, here it is *)
+  (* let run_in_pod node ~cmd =
+     (* run the cmd in every single container in a pod *)
+     let call_run container_id =
+       let%map res = run_in_container node ~container_id ~cmd in
+       res
+     in
+     let foldh accum m =
+       let open Deferred.Let_syntax in
+       let%bind acc = accum in
+       Deferred.map m ~f:(function s -> acc ^ s)
+     in
+     let container_list : string list =
+       List.append [ node.mina_container_id ]
+       (Option.to_list node.mina_archive_container_id)
+     in
+     let result_list = List.map container_list ~f:call_run in
+     List.fold result_list ~init:(Deferred.return "") ~f:foldh *)
 
   let start ~fresh_state node : unit Malleable_error.t =
     let open Malleable_error.Let_syntax in
     let%bind _ =
-      Deferred.bind ~f:Malleable_error.return (run_in_container node "ps aux")
+      Malleable_error.return
+        (run_in_container node ~container_id:node.mina_container_id
+           ~cmd:"ps aux")
     in
-    let%bind () =
+    let%bind _ =
       if fresh_state then
-        let%bind _ = run_in_container node "rm -rf .mina-config/*" in
-        Malleable_error.return ()
-      else Malleable_error.return ()
+        Malleable_error.return
+          (run_in_container node ~container_id:node.mina_container_id
+             ~cmd:"rm -rf .mina-config/*")
+      else Malleable_error.return (Deferred.return "")
+    in
+    let%bind _ =
+      Malleable_error.return
+        (run_in_container node ~container_id:node.mina_container_id
+           ~cmd:"/start.sh")
     in
-    let%bind _ = run_in_container node "/start.sh" in
     Malleable_error.return ()
 
   let stop node =
     let open Malleable_error.Let_syntax in
-    let%bind _ = run_in_container node "ps aux" in
-    let%bind _ = run_in_container node "/stop.sh" in
-    let%bind _ = run_in_container node "ps aux" in
+    let%bind _ =
```

Comment:
Also, just realized that the type here you are building in these binds is `'a Deferred.t Malleable_error.t` if I'm not mistaken. This doesn't work, and is unsafe, because what you have under the hood when we represent the computation is a `'a Deferred.t Deferred.t`. When you bind on that, you are not actually waiting for the inner deferred computation to happen, so you are creating race conditions since the code executing after the bind is not dependent on the previous computation finishing. You will want to go through and fix this up, as this mistake appears to be made throughout this PR and will lead to difficult to debug race conditions.

Reply:
To help avoid this, you can enforce that the left hand side of the `let%bind` must match a `unit` type. So, if you take both of my above suggestions along with this, the code would look something like:

```ocaml
let open Deferred.Or_error.Let_syntax in
let run_in_container' cmd =
  let%map _output : string = run_in_container node ~container_id:node.mina_container_id ~cmd in
  ()
in
let%bind () = run_in_container' "ps aux" in
let%bind () = run_in_container' "/stop.sh" in
let%bind () = run_in_container' "ps aux" in
Malleable_error.return ()  
```